### DISCUSSION TITLE
Native vim keybindings

### DISCUSSION THREAD

**Discussion by MartyMcFlyInTheSky** [zen-browser/desktop#2017] [Ideas]
I'm a huge fan of zen. Since I'm using vim keybindings all over the place, I'm asking if we could have native support for vim keybindings. Meaning, I plug a switch, and vim keybindings are automatically enabled for everything (though still modifiable of course!).

Now you're saying: Just use vimium or vimium-c, but these browser extensions have notable problems that seriously complicate and diminish the keyboard-only experience. Some of which are:
- No jumping back from browser address bar
- Some sites disable extensions for security reason (or more like zen/firefox disables the extensions). This means if I try to jump through tabs and I land on a browser-native page like `about:preferences#general` all the keybindings just stop to work and I have to reach for my mouse (or use the default keybindings of zen which I'm not very familiar with)
- Document editing sites like onedrive word/excel etc. will capture any keyboard input before the extension is able to do so, thus also creating an input "sink" where your flow completely dies.
- The extension is slow (javascript).

These are just some, I'm sure there more that people will comment on.

Therefore, zen could be the first browser to natively support vim keybindings like vimium and even go beyond it. We could add screen splitting keybindings and features comparable to nvim.telescope to zen. This would enhance the experience of powerusers a lot and also distinguish zen from all other browsers out there. Since I figure there are quite a lot of linux users among zen users (firefox squad), the target audience is covered. What do you think?

**Comment by ficd0** [comment]
As a vim addict myself, totally agreed. However, considering that Zen is (as far as I know) mostly developed by one person, it's unlikely for this to land at the top of the priorities list anytime soon. I would _love_ to see this feature too; but this feels like the sort of thing we as community members could consider implementing and making a PR for ourselves. I don't have any experience with browser development, but I feel passionate enough about this idea to learn. Let me know if you would be interested in pursuing this; I'm certain we can easily find even more passionate vim nerds to help!

**Reply by MartyMcFlyInTheSky** [reply]
I'm not into browsers at all but maybe I can take a look at the c++ part, now I'm kinda curious. Also huge respect for the sole developer working on this huge undertaking!

**Reply by ficd0** [reply]
absolutely. I actually had never heard of vimium before seeing your post. From the surface it seems amazing (not to invalidate your complaints with it; just saying the premise of using vim navigation in a browser seems really awesome!), so I'm going to install it and use it for a bit to really get a feel for it, and especially keeping an eye out for those complaints that you have so I can be wary of them and maybe brainstorm some solutions. 

If you want, you are welcome to email me at [daniel@ficd.ca](mailto:daniel@ficd.ca) if you want to discuss the topic further and maybe make some plans.

**Reply by MartyMcFlyInTheSky** [reply]
I like your enthusiasm @ficcdaf and I'd very much like to have the time at hand to drive this change. However atm I just don't have the time at hand. But I think there's no harm in risking a look and see where this leads us ;)

**Reply by ficd0** [reply]
similarly I'm very busy at the moment haha. I say re leave this thread here and see who else may also be interested in helping at some point.

**Comment by zebede22** [comment]
I would love to see the integration as well.
There is a Firefox extension which is superior to any other VIM emulators, maybe could be useful:
https://github.com/tridactyl/tridactyl

**Reply by neraid-live** [reply]
All due respect to Tridactyl, it is implemented as a browser _extension_, therefore it is limited to relying on an interface that is provided by the browser. Tridactyl actions are not avaliable before it is loaded onto the page, and as far as I'm aware there is no way to overcome that.

Having Vim-like motions avaliable as a _browser feature_ rather than an extension could greatly improve the navigation experience for some users.

**Comment by candsastle** [comment]
I totally agree with this request. So far I haven't found any browsers based off Firefox that can be used entirely with native vim-like bindings. I would love Zen Browser to be the first.

**Reply by harunnoir** [reply]
If this feature added to zen-browser it's gonna be the top browser ever!
